packages/flutter/lib/src/painting/_network_image_io.dart

```diff
@@ -81,39 +70,112 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+        } else if (response.error != null) {
+          bytesCompleter.completeError(response.error);
+        } else {
+          assert(false);
+        }
       });
-      final HttpClientResponse response = await request.close();
-      if (response.statusCode != HttpStatus.ok)
-        throw Exception('HTTP request failed, statusCode: ${response?.statusCode}, $resolved');
-
-      final Uint8List bytes = await consolidateHttpClientResponseBytes(
-        response,
-        onBytesReceived: (int cumulative, int total) {
-          chunkEvents.add(ImageChunkEvent(
-            cumulativeBytesLoaded: cumulative,
-            expectedTotalBytes: total,
-          ));
-        },
-      );
-      if (bytes.lengthInBytes == 0)
+
+      // This will keep reference to [debugNetworkImageHttpClientProvider] tree-shaken
+      // out of release builds.
+      HttpClientProvider httpClientProvider;
+      assert(() { httpClientProvider = debugNetworkImageHttpClientProvider; return true; }());
+
+      final _DownloadRequest downloadRequest = _DownloadRequest(downloadResponseHandler.sendPort,
+          resolved, headers, httpClientProvider);
+      if (_requestPort != null) {
+        // If worker isolate is properly set up ([_requestPort] is holding
+        // initialized [SendPort]), then just send download request down to it.
+        _requestPort.send(downloadRequest);
+      } else {
+        if (_pendingLoader == null) {
+          // If worker isolate creation was not started, start creation now.
+          _pendingLoadRequests = <_PendingLoadRequest>[];
+
+
+
+          _pendingLoader = _setupIsolate()..then((Isolate isolate) {
+              final RawReceivePort handleError = RawReceivePort((List<String> errorAndStackTrace) {
+                _cleanupDueToError(errorAndStackTrace[0]);
+              });
+              isolate.addErrorListener(handleError.sendPort);
+              isolate.resume(isolate.pauseCapability);
+            }).catchError((dynamic error, StackTrace stackTrace) {
+              _cleanupDueToError(error);
+            });
+        }
+        // Record download request so it can either send a request when isolate is ready or handle errors.
+        _pendingLoadRequests.add(_PendingLoadRequest(
+            (SendPort sendPort) { sendPort.send(downloadRequest); },
+            (dynamic error) { downloadRequest.sendPort.send(_DownloadResponse.error(error.toString())); }
+        ));
+      }
+
+      final TransferableTypedData transferable = await bytesCompleter.future;
+      downloadResponseHandler.close();
+
+      final Uint8List bytes = transferable.materialize().asUint8List();
+      if (bytes.isEmpty)
         throw Exception('NetworkImage is an empty file: $resolved');
 
-      return PaintingBinding.instance.instantiateImageCodec(bytes);
+      return await PaintingBinding.instance.instantiateImageCodec(bytes);
     } finally {
       chunkEvents.close();
     }
   }
 
+  void _cleanupDueToError(dynamic error) {
```

Comment:
Don't we need to `kill()` the isolate here?  The case I'm thinking of is when the isolate was successfully set-up, but then we get notified here via our `addErrorListener()` registration.

Reply:
Discussed offline